### PR TITLE
Esa gaia fix the cone_search_async to download json files

### PR DESCRIPTION
**Pull request by cosmoJFH**
The execution of the gaia astroquery function _cone_search_async_ throws the following error message when the user tries to retrieve the results in json format:
```
import astropy.units as u
from astropy.coordinates import SkyCoord
from astroquery.gaia import Gaia

dump_to_file= True

output_format='json'
output_file="my_file.json"


Gaia.ROW_LIMIT = 50  # Ensure the default row limit.
coord = SkyCoord(ra=280, dec=-60, unit=(u.degree, u.degree), frame='icrs')
j = Gaia.cone_search_async(coord, radius=u.Quantity(1.0, u.deg), output_file=output_file, output_format =output_format, dump_to_file=dump_to_file)

r = j.get_results()
r.pprint()


---------------------------------------------------------------------------
IORegistryError                           Traceback (most recent call last)
Cell In[26], line 15
     12 coord = SkyCoord(ra=280, dec=-60, unit=(u.degree, u.degree), frame='icrs')
     13 j = Gaia.cone_search_async(coord, radius=u.Quantity(1.0, u.deg), output_file=output_file, output_format =output_format, dump_to_file=dump_to_file)
---> 15 r = j.get_results()
     16 r.pprint()

File ~/anaconda3/lib/python3.10/site-packages/astroquery/utils/tap/model/job.py:235, in Job.get_results(self)
    231 # try load results from file
    232 # read_results_table_from_file checks whether
    233 # the file already exists or not
    234 outputFormat = self.parameters['format']
--> 235 results = modelutils.read_results_table_from_file(self.outputFile,
    236                                                   outputFormat)
    237 if results is not None:
    238     self.results = results

File ~/anaconda3/lib/python3.10/site-packages/astroquery/utils/tap/model/modelutils.py:39, in read_results_table_from_file(file_name, output_format, correct_units)
     35 astropy_format = utils.get_suitable_astropy_format(output_format)
     37 if check_file_exists(file_name):
---> 39     result = APTable.read(file_name, format=astropy_format)
     41     if correct_units:
     42         for cn in result.colnames:

File ~/anaconda3/lib/python3.10/site-packages/astropy/table/connect.py:62, in TableRead.__call__(self, *args, **kwargs)
     59 units = kwargs.pop("units", None)
     60 descriptions = kwargs.pop("descriptions", None)
---> 62 out = self.registry.read(cls, *args, **kwargs)
     64 # For some readers (e.g., ascii.ecsv), the returned `out` class is not
     65 # guaranteed to be the same as the desired output `cls`.  If so,
     66 # try coercing to desired class without copying (io.registry.read
     67 # would normally do a copy).  The normal case here is swapping
     68 # Table <=> QTable.
     69 if cls is not out.__class__:

File ~/anaconda3/lib/python3.10/site-packages/astropy/io/registry/core.py:220, in UnifiedInputRegistry.read(self, cls, format, cache, *args, **kwargs)
    217     if format == "votable":
    218         kwargs.update({"filename": path})
--> 220 reader = self.get_reader(format, cls)
    221 data = reader(*args, **kwargs)
    223 if not isinstance(data, cls):
    224     # User has read with a subclass where only the parent class is
    225     # registered.  This returns the parent class, so try coercing
    226     # to desired subclass.

File ~/anaconda3/lib/python3.10/site-packages/astropy/io/registry/core.py:153, in UnifiedInputRegistry.get_reader(self, data_format, data_class)
    151         return self._readers[(reader_format, reader_class)][0]
    152 format_table_str = self._get_format_table_str(data_class, "Read")
--> 153 raise IORegistryError(
    154     f"No reader defined for format '{data_format}' and class"
    155     f" '{data_class.__name__}'.\n\nThe available formats"
    156     f" are:\n\n{format_table_str}"
    157 )

IORegistryError: No reader defined for format 'json' and class 'Table'.

The available formats are:

           Format           Read Write Auto-identify Deprecated
--------------------------- ---- ----- ------------- ----------
                      ascii  Yes   Yes            No           
               ascii.aastex  Yes   Yes            No           
                ascii.basic  Yes   Yes            No           
                  ascii.cds  Yes    No            No           
     ascii.commented_header  Yes   Yes            No           
                  ascii.csv  Yes   Yes           Yes           
              ascii.daophot  Yes    No            No           
                 ascii.ecsv  Yes   Yes           Yes           
           ascii.fast_basic  Yes   Yes            No           
ascii.fast_commented_header  Yes   Yes            No           
             ascii.fast_csv  Yes   Yes            No           
       ascii.fast_no_header  Yes   Yes            No           
             ascii.fast_rdb  Yes   Yes            No           
             ascii.fast_tab  Yes   Yes            No           
          ascii.fixed_width  Yes   Yes            No           
ascii.fixed_width_no_header  Yes   Yes            No           
 ascii.fixed_width_two_line  Yes   Yes            No           
                 ascii.html  Yes   Yes           Yes           
                 ascii.ipac  Yes   Yes            No           
                ascii.latex  Yes   Yes           Yes           
                  ascii.mrt  Yes   Yes            No           
            ascii.no_header  Yes   Yes            No           
                  ascii.qdp  Yes   Yes           Yes           
                  ascii.rdb  Yes   Yes           Yes           
                  ascii.rst  Yes   Yes            No           
           ascii.sextractor  Yes    No            No           
                  ascii.tab  Yes   Yes            No           
                       fits  Yes   Yes           Yes           
                       hdf5  Yes   Yes           Yes           
                 pandas.csv  Yes   Yes            No           
                 pandas.fwf  Yes    No            No           
                pandas.html  Yes   Yes            No           
                pandas.json  Yes   Yes            No           
                    parquet  Yes   Yes           Yes           
                    votable  Yes   Yes           Yes           
                     aastex  Yes   Yes            No        Yes
                        cds  Yes    No            No        Yes
                        csv  Yes   Yes            No        Yes
                    daophot  Yes    No            No        Yes
                       html  Yes   Yes            No        Yes
                       ipac  Yes   Yes            No        Yes
                      latex  Yes   Yes            No        Yes
                        mrt  Yes   Yes            No        Yes
                        rdb  Yes   Yes            No        Yes
```

This issue is present in the functions

1. cone_search
2. cone_search_async
3. launch_job
4. launch_job_async

cc @esdc-esac-esa-int

Jira ticket GAIAPCR-1302

### CODE REVIEW

**Comment by cosmoJFH** [comment]
The following solution has been suggested (based on the json file returned by the gaia system). Gaia returns a json made of 2 "blocks": tthe first one, where he  key  _metadata_ that is associated to a list of dictionaries, that provides additional information for the data, and the second one, where the key _data_ is associated to a list of lists, that contains  the data. 

It looks like that the returned json file cannot be handled by the Table, since each key has a different structured associated to it. So we have split the  original json in two: one for the metadata and the other for the data. They are returned to the user. The original json file is saved to the file system is the user defines the parameters output_file and dump_to_file for the output_format='json'.

File: astroquery/utils/tap/model/modelutils.py [hunk 1]
```diff
@@ -36,11 +37,25 @@ def read_results_table_from_file(file_name, output_format, *, correct_units=True
...
+                    files['data'] = APTable.read(json.dumps({"data": data['data']}), format=astropy_format)
+                    files['metadata'] = APTable.read(json.dumps({"metadata": data['metadata']}), format=astropy_format)
+
+                    return files
+                else:
+                    return APTable.read(file_name, format=astropy_format)
+        else:
+            result = APTable.read(file_name, format=astropy_format)
+
+            if correct_units:
```
Comment (bsipocz) [inline astroquery/utils/tap/model/modelutils.py:56]: are units correct for the JSON output?

**Review by bsipocz — CHANGES_REQUESTED** [review]
It would be good to include this in the tests, either in the gaia tests or for the tap ones.

**Comment by jespinosaar** [comment]
> It would be good to include this in the tests, either in the gaia tests or for the tap ones.

Totally agree, we should always include some tests with the new code. @cosmoJFH , would you mind to include it? Thanks!

File: astroquery/utils/tap/model/modelutils.py (hunk 1 above)
Comment (cosmoJFH) [reply astroquery/utils/tap/model/modelutils.py:70]: Hi @bsipocz, in order to correct the units in the astropy Table, we have changed the implementation: a Table is build from scratch based on the returned json file and the units are corrected. The original json is directly saved into the file system if the parameters dump_to_file= True and output_file are set.

**Comment by cosmoJFH** [comment]
> > It would be good to include this in the tests, either in the gaia tests or for the tap ones.
> 
> Totally agree, we should always include some tests with the new code. @cosmoJFH , would you mind to include it? Thanks!

Hi @jespinosaar, we have made new tests in the gaia project.

**Comment by cosmoJFH** [comment]
A part from the described changes, the following tasks were also completed for the method  `load_data`, 

1. the code was updated to process files for the format ecsv.
3. the format json is removed from the documentation, since the gaia system does not generate ancillary data with this format.
4. the extension '.zip' is added to the value of the parameter output_file in case it does not contain it, since the data is always downloaded as a zipped file.

**Comment by cosmoJFH** [comment]
hi @bsipocz, do you know why the check "CI-devtest / dev dependencies with all dependencies with coverage (pull_request) " is failing? All the error messages are like this one: 

- DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)


Should I do anything to fix this?

**Comment by bsipocz** [comment]
@cosmoJFH - that failure is unrelated and was due to astropy dev version incompatibility with numpy 2.0. It has been fixed, but we started to see other issues, too, that also has been addressed, and now a third one. 

Long story short, don't worry about that job for now.

**Review by bsipocz — APPROVED** [review]
There are significant code repetition in the new tests that I think could be cut back with using parametrization for pytest. But it's really not a big deal to leave it as is as it's only for the tests, so let me know if you rather not refactor them.

File: astroquery/gaia/core.py [hunk 2]
```diff
@@ -236,6 +238,14 @@ def load_data(self, ids, *, data_release=None, data_structure='INDIVIDUAL', retr
             output_file = os.path.join(os.getcwd(), temp_dirname, downloadname_formated)
         else:
             output_file_specified = True
+
+            if isinstance(output_file, str):
+                if not output_file.lower().endswith('.zip'):
+                    output_file = output_file + '.zip'
+            elif isinstance(output_file, Path):
```
[the remark this answers is not in the document]
Comment (cosmoJFH) [reply astroquery/gaia/core.py:245]: The docstring has been updated.

**Comment by bsipocz** [comment]
Thanks @cosmoJFH!

**Pull request merged**